fog/view/enclave/impl/src/types.rs

```diff
@@ -0,0 +1,327 @@
+                query_response,
+                block_range,
+            };
+            decrypted_query_responses.push(decrypted_query_response);
+        }
+
+        let shared_data: SharedData = decrypted_query_responses.as_slice().into();
+
+        let actual_missed_block_ranges =
+            HashSet::<_>::from_iter(shared_data.missed_block_ranges.iter());
+        let actual_rng_records = HashSet::<_>::from_iter(shared_data.rng_records.iter());
+        let actual_decommissioned_ingest_invocations =
+            HashSet::<_>::from_iter(shared_data.decommissioned_ingest_invocations.iter());
+
+        let expected_missed_block_ranges = HashSet::<_>::from_iter(missed_block_ranges.iter());
+        let expected_rng_records = HashSet::<_>::from_iter(rng_records.iter());
+        let expected_decommissioned_ingest_invocations =
+            HashSet::<_>::from_iter(decommissioned_ingest_invocations.iter());
+
+        assert_eq!(actual_missed_block_ranges, expected_missed_block_ranges);
+        assert_eq!(actual_rng_records, expected_rng_records);
+        assert_eq!(
+            actual_decommissioned_ingest_invocations,
+            expected_decommissioned_ingest_invocations
+        );
+        assert_eq!(
+            shared_data.next_start_from_user_event_id,
+            NEXT_START_FROM_USER_EVENT_ID
+        );
+    }
+
+    #[test]
+    fn responses_have_different_values() {
+        const STORE_COUNT: usize = 4;
+        let mut decrypted_query_responses = Vec::with_capacity(STORE_COUNT);
+
+        let missed_block_ranges = vec![
+            BlockRange::new(0, 1),
+            BlockRange::new(10, 12),
+            BlockRange::new(33, 100),
+            BlockRange::new(100, 200),
+        ];
+
+        let mut rng_records = Vec::with_capacity(STORE_COUNT);
+        for i in 0..STORE_COUNT {
+            let egress_public_key = KexRngPubkey {
+                public_key: vec![i as u8; 32],
+                version: i as u32,
+            };
+            let rng_record = RngRecord {
+                ingest_invocation_id: i as i64,
+                pubkey: egress_public_key,
+                start_block: 0,
+            };
+            rng_records.push(rng_record);
+        }
+
+        let mut decommissioned_ingest_invocations = Vec::with_capacity(STORE_COUNT);
+        for i in 0..STORE_COUNT {
+            let decommissioned_ingest_invocation = DecommissionedIngestInvocation {
+                ingest_invocation_id: i as i64,
+                last_ingested_block: 10,
+            };
+            decommissioned_ingest_invocations.push(decommissioned_ingest_invocation);
+        }
+
+        for i in 0..STORE_COUNT {
+            let missed_block_ranges = vec![missed_block_ranges[i].clone()];
+            let rng_records = vec![rng_records[i].clone()];
+            let decommissioned_ingest_invocations =
+                vec![decommissioned_ingest_invocations[i].clone()];
```

Comment:
:wrench: I'm not fond of the `+1` everywhere in the suggestion, but it felt like the deduping need/logic wasn't being exercised.
```suggestion
        let missed_block_ranges = vec![
            BlockRange::new(0, 1),
            BlockRange::new(10, 12),
            BlockRange::new(33, 100),
            BlockRange::new(100, 200),
            BlockRange::new(205, 210),
        ];

        let mut rng_records = Vec::with_capacity(STORE_COUNT + 1);
        for i in 0..=STORE_COUNT {
            let egress_public_key = KexRngPubkey {
                public_key: vec![i as u8; 32],
                version: i as u32,
            };
            let rng_record = RngRecord {
                ingest_invocation_id: i as i64,
                pubkey: egress_public_key,
                start_block: 0,
            };
            rng_records.push(rng_record);
        }

        let mut decommissioned_ingest_invocations = Vec::with_capacity(STORE_COUNT + 1);
        for i in 0..=STORE_COUNT {
            let decommissioned_ingest_invocation = DecommissionedIngestInvocation {
                ingest_invocation_id: i as i64,
                last_ingested_block: 10,
            };
            decommissioned_ingest_invocations.push(decommissioned_ingest_invocation);
        }

        for i in 0..STORE_COUNT {
            let missed_block_ranges = vec![missed_block_ranges[i].clone(), missed_block_ranges[i + 1].clone()];
            let rng_records = vec![rng_records[i].clone(), rng_records[i + 1].clone()];
            let decommissioned_ingest_invocations =
                vec![decommissioned_ingest_invocations[i].clone(), decommissioned_ingest_invocations[i + 1].clone()];
```

Reply:
The dedup behavior is being tested in the previous test method `responses_have_same_values`. I want this test case to test when there is zero overlap between the different shards, so I'll keep it as is.

Reply:
oh yah...  Don't know what I was thinking